### PR TITLE
Configurable session timeouts

### LINKED ISSUE #1939: Configurable session timeouts
**Is your feature request related to a problem? Please describe.**

Currently, session timeouts are hard-coded at compile time, 3600 seconds (1h) for a normal token, 600s (10min) for "elevated" or write access.

**Describe the solution you'd like**

Allow it to be configurable at a stystem level, changeable by admin-level users, stored as a value in the database.

**Describe alternatives you've considered**

Allow it to be set in the server config file

**Additional context**

### PR DESCRIPTION
**Pull request by Seba-T**
Fixes #1939

Checklist

- [x] This pr contains no AI generated code
- [x] cargo fmt has been run
- [x] cargo clippy has been run
- [x] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: server/lib/src/server/mod.rs [hunk 1]
```diff
@@ -818,6 +818,36 @@ pub trait QueryServerTransaction<'a> {
...
+        self.internal_search_uuid(UUID_SYSTEM_CONFIG)
+            .and_then(|e| {
+                if let Some(expiry_time) = e.get_ava_single_uint32("authsession_expiry") {
+                    Ok(expiry_time)
+                } else {
+                    Err(OperationError::NoMatchingAttributes)
+                }
+            })
+            .map_err(|e| {
+                admin_error!(?e, "Failed to retrieve system configuration");
```
Comment (yaleman) [inline server/lib/src/server/mod.rs:831]: smol nitpick

```suggestion
                admin_error!(?e, "Failed to retrieve authsession_expiry from system configuration");
```

File: server/lib/src/server/mod.rs [hunk 2]
```diff
@@ -818,6 +818,36 @@ pub trait QueryServerTransaction<'a> {
...
+        self.internal_search_uuid(UUID_SYSTEM_CONFIG)
+            .and_then(|e| {
+                if let Some(expiry_time) = e.get_ava_single_uint32("privilege_expiry") {
+                    Ok(expiry_time)
+                } else {
+                    Err(OperationError::NoMatchingAttributes)
+                }
+            })
+            .map_err(|e| {
+                admin_error!(?e, "Failed to retrieve system configuration");
```
Comment (yaleman) [inline server/lib/src/server/mod.rs:846]: smol nitpick

```suggestion
                admin_error!(?e, "Failed to retrieve privilege_expiry from system configuration");
```

**Review by yaleman — CHANGES_REQUESTED** [review]
Minor nits picked but otherwise a solid PR 👍🏻

**Comment by Seba-T** [comment]
I've tried to fix the tests as I've realized that what I had written made no sense but so far nothing I'm trying is actually working. Should I push it anyway so maybe I can better showcase what the issue is?

File: libs/client/src/system.rs [hunk 3]
```diff
@@ -23,4 +23,35 @@ impl KanidmClient {
         self.perform_delete_request_with_body("/v1/system/_attr/badlist_password", list)
             .await
     }
+    pub async fn system_authsession_expiry_get(&self) -> Result<u32, ClientError> {
+        let list: Option<[String; 1]> = self
+            .perform_get_request("/v1/system/_attr/authsession_expiry")
+            .await?;
+        list.ok_or(ClientError::EmptyResponse)
+            .map(|s| s[0].parse::<u32>().unwrap_or_default())
```
[the remark this answers is not in the document]
Comment (yaleman) [reply libs/client/src/system.rs:31]: Oh, client lib, that's my derp. Just use the logging `error!()` which should be in the crate

**Comment by yaleman** [comment]
> I've tried to fix the tests as I've realized that what I had written made no sense but so far nothing I'm trying is actually working. Should I push it anyway so maybe I can better showcase what the issue is?

Yep, push it and we can look, for sure

**Comment by Seba-T** [comment]
Basically the issue is that all the ways I came up to test the expiry time always ended up having a `SessionScope::PrivilegeCapable` that has a fixed expiry of 24 hrs, so it doesn't matter what I set the expiry to. Furthermore I was mislead by the previous test that has this piece of code:
```
let post_grace = ct + GRACE_WINDOW + Duration::from_secs(1);
let expiry = ct + Duration::from_secs(DEFAULT_AUTH_SESSION_EXPIRY as u64 + 1);

// Assert that our grace time is less than expiry, so we know the failure is due to
// this.
assert!(post_grace < expiry);
```

But this actually makes no sense as the expiry time for the admin session is always 24hrs so I don't know what William was thinking when he wrote it. Should I open an issue about it?. 
So basically I just need a way to crate a non privileged session and then everything should be fine

File: libs/client/src/system.rs [hunk 4]
```diff
@@ -23,4 +23,39 @@ impl KanidmClient {
         self.perform_delete_request_with_body("/v1/system/_attr/badlist_password", list)
             .await
     }
+    pub async fn system_authsession_expiry_get(&self) -> Result<u32, ClientError> {
```
Comment (Firstyear) [inline libs/client/src/system.rs:26]: space above this line.

File: server/lib/src/constants/acp.rs [hunk 5]
```diff
@@ -1291,7 +1291,14 @@ lazy_static! {
         ("acp_search_attr", Value::new_iutf8("description")),
         ("acp_search_attr", Value::new_iutf8("badlist_password")),
         ("acp_modify_removedattr", Value::new_iutf8("badlist_password")),
-        ("acp_modify_presentattr", Value::new_iutf8("badlist_password"))
+        ("acp_modify_presentattr", Value::new_iutf8("badlist_password")),
```
Comment (Firstyear) [inline server/lib/src/constants/acp.rs:1294]: Should probably be a seperate ACP for IDM_ACP_SYSTEM_CONFIG_SESSION_EXP, and then we tie it to the group member instead? That matches in with the other future changes to fine-grain acp later.

File: server/lib/src/idm/account.rs [hunk 6]
```diff
@@ -542,6 +545,7 @@ impl Account {
 
         // Anonymous does NOT record it's sessions, so we simply check the expiry time
         // of the token. This is already done for us as noted above.
+        dbg!(uat.clone());
```
Comment (Firstyear) [inline server/lib/src/idm/account.rs:548]: debug/trace rather than dbg in production code here.

File: server/lib/src/idm/authsession.rs [hunk 7]
```diff
@@ -1007,7 +1008,10 @@ impl AuthSession {
         webauthn: &Webauthn,
         pw_badlist_set: Option<&HashSet<String>>,
         uat_jwt_signer: &JwsSigner,
+        auth_session_expiry: u32,
+        auth_privilege_expiry: u32,
     ) -> Result<AuthState, OperationError> {
+        dbg!(auth_session_expiry, auth_privilege_expiry);
```
Comment (Firstyear) [inline server/lib/src/idm/authsession.rs:1014]: trace or debug.

File: server/lib/src/idm/server.rs [hunk 8]
```diff
@@ -90,6 +90,8 @@ pub struct IdmServer {
     pw_badlist_cache: Arc<CowCell<HashSet<String>>>,
     oauth2rs: Arc<Oauth2ResourceServers>,
     domain_keys: Arc<CowCell<DomainKeys>>,
+    privilege_expiry: Arc<CowCell<u32>>,
```
Comment (Firstyear) [inline server/lib/src/idm/server.rs:93]: I think rather then a pair of u32, this should be "default_account_policy" and a struct of "AccountPolicy" which contains the u32s.) This is because we'll need to extend this for more policies later, and this "fan out" of cowcell does have a minor cost at run time that we want to somewhat avoid. This would also let you do "AccountPolicy::default" in the validate cred step and prevent too many arcs in fn calls, and make it easier to extend account policy later. As well, it'll let us later, overlay and combine the account policy from the system and the user's resulting set of policy once we add that.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: server/lib/src/idm/server.rs (hunk 8 above)
Comment (Seba-T) [reply server/lib/src/idm/server.rs:93]: should we also move the pw badlist cache inside this `AccountPolicy`?

File: server/lib/src/idm/account.rs (hunk 6 above)
Comment (Seba-T) [reply server/lib/src/idm/account.rs:548]: I was just fighting against failing tests, I didn't mean to leave it in prod :p

File: server/lib/src/constants/acp.rs (hunk 5 above)
Comment (Seba-T) [reply server/lib/src/constants/acp.rs:1294]: sure!

File: server/lib/src/idm/server.rs (hunk 8 above)
Comment (Firstyear) [reply server/lib/src/idm/server.rs:93]: Actually yes! This is an EXCELLENT idea :)

File: server/lib/src/idm/account.rs (hunk 6 above)
Comment (Firstyear) [reply server/lib/src/idm/account.rs:548]: Naughty naughty :P  It's why trace! is good here since that's meants to be developer-level logs we can accidentally leave in. We plan to do a big overhaul of log messages soon to make sure they are all at proper levels and details.

File: libs/client/src/system.rs (hunk 4 above)
Comment (Firstyear) [reply libs/client/src/system.rs:26]: I feel like you do this just to see if I'm paying attention :P

File: server/lib/src/idm/server.rs (hunk 8 above)
Comment (Seba-T) [reply server/lib/src/idm/server.rs:93]: awesome!

**Comment by Firstyear** [comment]
Also remember to rebase on latest master due to all the recent changes :)

File: libs/client/src/system.rs (hunk 4 above)
Comment (Seba-T) [reply libs/client/src/system.rs:26]: Dang you got me /s

**Comment by Seba-T** [comment]
Yeah I'm going to get on this after the identity stuff is done once for all

**Comment by Seba-T** [comment]
Alright I think I fixed everything and I also got the test working!!

File: server/lib/src/idm/server.rs [hunk 9]
```diff
@@ -3589,6 +3658,137 @@ mod tests {
...
+            .await
+            .validate_and_parse_token_to_token(Some(&token), ct)
+            .expect("Must not fail") else {
+            panic!("Unexpected auth token type for anonymous auth");
+            };
+
+        debug!(?uat);
+
+        assert!(
+            matches!(uat.expiry, Some(x) if dbg!(x) == OffsetDateTime::UNIX_EPOCH + ct + Duration::from_secs(new_authsession_expiry as u64))
```
Comment (Firstyear) [inline server/lib/src/idm/server.rs:3788]: What does this dbg(x) here do?

File: server/lib/src/idm/server.rs [hunk 10]
```diff
@@ -1979,7 +2024,11 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
             .get_changed_uuids()
             .contains(&UUID_SYSTEM_CONFIG)
         {
+            // TODO: probably it would be more efficient to introduce a single check for each of the possible system configs
+            // but that would require changing what uuid the operation is assigned
             self.reload_password_badlist()?;
+            self.reload_authsession_expiry()?;
```
Comment (Firstyear) [inline server/lib/src/idm/server.rs:2030]: Could these all be coallesced into a single "reload_system_account_policy"?

**Review by Firstyear — CHANGES_REQUESTED** [review]
Some minor changes/cleanups but looks super good! Well done. I think you forgot to add the admin interfaces to the cli though so admins can actually mod/set these values.

File: server/lib/src/idm/server.rs (hunk 10 above)
Comment (Seba-T) [reply server/lib/src/idm/server.rs:2030]: Actually, yes 😄

File: server/lib/src/idm/server.rs (hunk 9 above)
Comment (Seba-T) [reply server/lib/src/idm/server.rs:3788]: Add more pollution to your terminal when you run the test :p. It’s just a leftover from when I was fightings against the test, I’ll remove it

**Comment by Seba-T** [comment]
> Some minor changes/cleanups but looks super good! Well done. I think you forgot to add the admin interfaces to the cli though so admins can actually mod/set these values.

Done! :smile:

**Review by Firstyear — APPROVED** [review]

**Comment by Firstyear** [comment]
@yaleman Needs your tick of approval :)

File: server/lib/src/constants/schema.rs [hunk 11]
```diff
@@ -193,6 +193,24 @@ pub static ref SCHEMA_ATTR_BADLIST_PASSWORD: SchemaAttribute = SchemaAttribute {
     ..Default::default()
 };
 
+pub static ref SCHEMA_ATTR_AUTH_SESSION_EXPIRY: SchemaAttribute = SchemaAttribute {
+    uuid: UUID_SCHEMA_ATTR_AUTH_SESSION_EXPIRY,
+    name: "authsession_expiry".into(),
```
Comment (yaleman) [inline server/lib/src/constants/schema.rs:198]: This should use the `enum Attribute.into()`

File: server/lib/src/constants/schema.rs [hunk 12]
```diff
@@ -193,6 +193,24 @@ pub static ref SCHEMA_ATTR_BADLIST_PASSWORD: SchemaAttribute = SchemaAttribute {
...
+
+    description: "An expiration time for an authentication session.".to_string(),
+    syntax: SyntaxType::Uint32,
+    ..Default::default()
+};
+
+pub static ref SCHEMA_ATTR_AUTH_PRIVILEGE_EXPIRY: SchemaAttribute = SchemaAttribute {
+    uuid: UUID_SCHEMA_ATTR_AUTH_PRIVILEGE_EXPIRY,
+    name: "privilege_expiry".into(),
```
Comment (yaleman) [inline server/lib/src/constants/schema.rs:207]: This should use the `enum Attribute.into()`

**Review by yaleman — CHANGES_REQUESTED** [review]
Minor changes needed, then good to go

File: server/lib/src/constants/schema.rs (hunk 12 above)
Comment (Seba-T) [reply server/lib/src/constants/schema.rs]: Done :)

**Review by yaleman — APPROVED** [review]
